### DISCUSSION TITLE
Applying compositing operations to multiple images in a stack

### DISCUSSION THREAD

**Discussion by dpthurst** [ImageMagick/ImageMagick#5888] [Help] [answered]
I have a multi-page PDF file with card images, and I would like to perform various operations on the entire stack: change the resolution, trim some from the sides, round the corners, and export out to a sequence of images. The first few operations are 'simple image processing operations' and apply to all the images in memory, as I want. But rounding the corners is most conveniently done with the "-composite" operation, which operates only on the last two images in the stack.

Is there a way to do this conveniently in ImageMagick? Obviously I can write out the intermediate images to files and operate on them individually, but it seems like an extra annoyance. I note that the lack of such operations appears other places, notably in the '-annotate' command which does one particular inflexible instance of such an operation.

Here's the command I'm trying:
`magick -density 300 -colorspace sRGB t.pdf[657x897] -crop 600x840+28+28 \( -size 600x840 xc:none -draw "roundrectangle 0,0,599,839,28,28" \) -compose DstIn -composite %02d.jpg`
and here's the output if I give it a 1-page PDF:
![00](https://user-images.githubusercontent.com/13771964/208315749-feb90194-58be-4fb1-a73c-8a91ed1c6097.jpg)
(This is mostly what I want, except that the corners are transparent, but I can't upload the WEBP output which would be my preferred output format; PNGs are too large.)

If I give it a PDF with 2 or more pages as input, all of them get flattened, and the corners get understandably messed up. I unfortunately can't upload the original PDFs I'm using as input.

**Comment by snibgo** [comment, marked as answer]
I'm not sure what question(s) you are asking.

Your input PDF contains raster images. When IM reads the PDF, it rasterizes each PDF page. This re-samples each embedded raster image. You probably get better quality by using `pdfimages` to extract each embedded image without resampling.

When you have a number of input images and want to compose each one individually with some other fixed image, you can use "`-layers composite`" with "`null:`" between them, like this:
```
magick \
  in1.png in2.png in3.png in4.png \
  -crop 600x840+28+28 ^
  NULL: \
  \( -size 600x840 xc:none -draw "roundrectangle 0,0,599,839,28,28" \) \
  -compose DstIn -layers composite \
  out_%02d.png
```
Does that help?

**Reply by dpthurst** [reply]
Thank you!

> Your input PDF contains raster images. When IM reads the PDF, it rasterizes each PDF page. This re-samples each embedded raster image. You probably get better quality by using pdfimages to extract each embedded image without resampling.

I'm aware of this, of course. Unfortunately the embedded images are overlaid in a complicated way, and it is too hard to work with the pdfimages output. They are at 300dpi (hence that command line option, you'll notice the output image is at lower than 300dpi), though I believe I might still get better quality by doing intermediate processing at a higher resolution; that is next on my agenda to sort out.

> When you have a number of input images and want to compose each one individually with some other fixed image, you can use "-layers composite" with "null:" between them, like this: ...

That is almost what I wanted, but ImageMagick has a virtual canvas that it treats differently with the `-composite` operation I tried first and the `-layers composite` operation you suggest. So I believe I need to add a `+repage` operation as follows:

```
magick \
  in1.png in2.png in3.png in4.png \
  -crop 600x840+28+28 +repage \
  NULL: \
  \( -size 600x840 xc:none -draw "roundrectangle 0,0,599,839,28,28" \) \
  -compose DstIn -layers composite \
  out_%02d.png
```

**Comment by snibgo** [comment]
>  So I believe I need to add a +repage operation ...

Yes, I didn't notice that, sorry.
